File: leetcode/29.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <sstream>

using namespace std;

class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == 0) return 0;

        if (dividend == INT_MIN) {
            if (divisor == 1) {
                return INT_MIN;
            }

            if (divisor == -1) {
                return INT_MAX;
            }
        }

        bool neg = (dividend < 0) ^ (divisor < 0);
        int result = 0;
        long long ldividend = abs(dividend);
        long long ldivisor = abs(divisor);

        cout << ldividend << ' ' << ldivisor << endl;

        while (ldividend >= ldivisor) {
            ldividend -= ldivisor;
            result++;
        }

        if (neg) {
            return ~result + 1;
        } else {
            return result;
        }
    }
};

int main() {
    Solution s;
    
    int r = s.divide(INT_MIN, 1);

    cout << r << endl;
}

File: leetcode/29.js
/**
 * @param {number} dividend
 * @param {number} divisor
 * @return {number}
 */
var divide = function(dividend, divisor) {
    const lowerBound = -(1<<31>>>0);
    const upperBound = (1<<31>>>0) - 1;
    const polar = (dividend ^ divisor) >= 0;
    const absDividend = Math.abs(dividend);
    const absDivisor = Math.abs(divisor);

    const absResult = Math.floor(Math.exp(Math.log(absDividend) - Math.log(absDivisor)));

    const result = polar ? absResult : -absResult;

    if (result < lowerBound || result > upperBound) {
        return upperBound;
    }

    return result;
};

const lowerBound = -(1<<31>>>0);
const upperBound = 1<<31>>>0 - 1;

const cases = [
    [0, 1],
    [1, 1],
    [1, -1],
    [-1, 1],
    [-1, -1],
    [lowerBound, 1],
    [lowerBound, -1]
];

cases.forEach(([a, b]) => console.log(divide(a, b)));
